### PR TITLE
MODIT+VALD

### PR DESCRIPTION
**Pull request by chonma0ctopus**
Enable to compute the atomic (& ionic) lines from VALD with MODIT.
The database of VALD lines is automatically divided into different species and after the optical depth of each species is calculated, they are summed up.

### CODE REVIEW

**Comment by HajimeKawahara** [comment]
@chonma0ctopus Thanks! 
When I run the unit test, 
```sh
shirochan:~/exojax/tests(modit-vald)>python vald_modit/VALD_MODIT_test.py
```
it fails probably in jax.lax.scan.  Can you send your version of JAX?

```sh
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "vald_modit/VALD_MODIT_test.py", line 176, in <module>
    test_VALD_MODIT()
  File "vald_modit/VALD_MODIT_test.py", line 85, in test_VALD_MODIT
    dgm_ngammaL_VALD = setdgm_vald_all(asdb, PH_ref, PHe_ref, PHH_ref, R, fT, res, T0_test, alpha_test)
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/exojax-1.0.0-py3.8.egg/exojax/spec/modit.py", line 446, in setdgm_vald_all
    dgm_ngammaL_sp = setdgm_vald_each(asdb.ielem[i], asdb.iion[i], asdb.atomicmass[i], asdb.ionE[i], \
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/exojax-1.0.0-py3.8.egg/exojax/spec/modit.py", line 411, in setdgm_vald_each
    ngammaLM = scan(floop, 0, ngammaLM)[1]
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/exojax-1.0.0-py3.8.egg/exojax/spec/modit.py", line 410, in <lambda>
    floop = lambda c, arr:  (c, jnp.nan_to_num(arr, nan=jnp.nanmin(arr)))
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 2052, in nan_to_num
    x = where(isnan(x), _constant_like(x, nan), x)
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 379, in _constant_like
    return np.array(const, dtype=_dtype(x))
jax._src.errors.TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=1/0)>
While tracing the function <lambda> at /home/kawahara/anaconda3/lib/python3.8/site-packages/exojax-1.0.0-py3.8.egg/exojax/spec/modit.py:410 for scan, this concrete value was not available in Python because it depends on the value of the argument 'arr'.
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError
```

**Comment by HajimeKawahara** [comment]
my environment: 
jax                                0.2.19
jaxlib                             0.1.70+cuda111

**Comment by chonma0ctopus** [comment]
Thank you for checking.
My environment:
jax=0.2.22=pypi_0
jaxlib=0.1.72+cuda111=pypi_0

**Comment by HajimeKawahara** [comment]
Thanks. This looks we need newer version of JAX >= 0.2.22?

This environment also passed the unit test. 
jax                                0.3.2
jaxlib                             0.3.0+cuda11.cudnn82

**Comment by HajimeKawahara** [comment]
I suspect [this revision, jax 0.2.21](https://github.com/google/jax/releases/tag/jax-v0.2.21) about conversion of ndarray makes difference.  They say:

**Comment by chonma0ctopus** [comment]
I see. Is it ok to recommend users to run in the newer version than the 0.2.21?

**Comment by HajimeKawahara** [comment]
I think so, but not confirmed for v0.2.21. So probably >=v0.2.22?

**Comment by chonma0ctopus** [comment]
Understood. 
So, where can I add the recommendation message? README.md or somewhere else?

**Comment by HajimeKawahara** [comment]
I think we should directly specify in `setup.py`.

**Comment by HajimeKawahara** [comment]
a bit large difference prevented to pass the unit test for `xs/opacity_Fe_test.py`:
Can you check it in your environment? @chonma0ctopus 

```sh
manbou ~/exojax/tests(modit-vald)>python xs/opacity_Fe_test.py
Reading VALD file
[9.30509769e-21 1.25000001e-05 3.92851562e-03]
Traceback (most recent call last):
  File "xs/opacity_Fe_test.py", line 174, in <module>
    test_opacity_Fe_vald3(2995., 0.1)
  File "xs/opacity_Fe_test.py", line 84, in test_opacity_Fe_vald3
    assert(diff[0] < 1.e-11 and diff[1] < 1.e-3 and diff[2] < 1.e-3)
AssertionError
```

**Comment by chonma0ctopus** [comment]
Is this correct?

INSTALL_REQUIRES = ['numpy', 'pandas', 'tqdm', 'scipy',
                    'jax>=v0.2.22', 'numpyro', 'pyarrow', 'bs4', 'vaex',"jaxopt"]

**Comment by chonma0ctopus** [comment]
> a bit large difference prevented to pass the unit test for `xs/opacity_Fe_test.py`: Can you check it in your environment? @chonma0ctopus

Thank you for finding out. I'll check it.

**Comment by HajimeKawahara** [comment]
I think
`jax>=0.2.22`
remove `v`

**Comment by chonma0ctopus** [comment]
True!💦 Thank you.

**Comment by chonma0ctopus** [comment]
> a bit large difference prevented to pass the unit test for `xs/opacity_Fe_test.py`: Can you check it in your environment? @chonma0ctopus
> 
I found that this is caused due to the change of a procedure to extract ionization energy (adb.ionE) in class AdbVald. (lines 896–898 of moldb.py)
It is revised to use NIST data of ionization energy (NIST_Atomic_Ionization_Energies.txt).
They have higher precision than the values in atomic.txt, which we previously used.

So, I have updated the REFS in opacity_Fe_test.py according to this change.

**Comment by HajimeKawahara** [comment]
confirmed. thanks.

**Comment by HajimeKawahara** [comment]
@chonma0ctopus 
Can you update [the user guide for atomll](https://github.com/chonma0ctopus/exojax/blob/modit-vald/documents/userguide/atomll.rst)? if needed.

File: src/exojax/spec/atomll.py [hunk 1]
```diff
@@ -384,3 +390,206 @@ def gamma_KA3s(T, PH, PHH, PHe, ielem, iion,
...
+        return msi[0]
+    
+    mods_uspecies_list = jnp.zeros(len(uspecies))
+    mods_uspecies_list = g_Mmul([mods_uspecies_list, 0])
+
+    return (MMR_uspecies_list, atomicmass_uspecies_list, mods_uspecies_list)
+    
+    
+def sep_arr_of_sp(arr, adb, trans_jnp=True, inttype=False):
+    """Split by species the jnp.array stored as instance variable in adb, and pad with zeros to adjust the length
```
Comment (HajimeKawahara) [inline src/exojax/spec/atomll.py:533]: Please add Args and Returns.

**Comment by HajimeKawahara** [comment]
I confirmed that Reverse modeling with VALD using MODIT works in my laptop. Nice work! @chonma0ctopus 

Is this the world's first HMC-fitted M-dwarf spectrum?
gj699
!

**Review by HajimeKawahara — APPROVED** [review]
I approve the current branch.

**Pull request merged**